e2e: upgrade: tests for topology manager

ffromani opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
This is another followup of #86184 . I'm working with @vpickard to add more e2e tests for the topology manager to fix the issue #83481

**Which issue(s) this PR fixes**:
Related to #83481

**Special notes for your reviewer**:
N/A

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

ffromani committed:
e2e: upgrade: tests for topology manager

This patch add the first upgrade test for the topology manager.
At the moment of writing, the topology manager is graduating to beta,
and we want to check it keeps being functional across upgrades.

A number of tradeoffs are made in this patch:
1. the test intentionally does not try to enable the topology manager on
   worker nodes. This is to keep the test simple and to avoid it to be
   Disruptive.

2. for the same simplicity reasons, the test doesn't check the worker
   nodes are actually multi-NUMA machines. This is left for the test
   environment operators.

2. the test intentionally only check the minimum alignment requisites,
   the CPU cores. Intentionally avoid to require devices in this patch
   to make the test easier to run - less prerequisites.

We believe that even considering these tradeoffs the test can deliver
a good baseline signal of the topology manager functionality across
upgrades.

Signed-off-by: Francesco Romani <fromani@redhat.com>

fromanirh commented:
/hold
still WIP

fromanirh commented on test/e2e/upgrades/topology_manager.go:
```diff
@@ -0,0 +1,317 @@
+	return nil
+}
+
+// TODO: this dupes test/e2e/windows/memory_limits.go. How can we factor this code out?
```
I have another PR in the working to fix this

fromanirh replied:
Here: https://github.com/kubernetes/kubernetes/pull/88110

fromanirh commented:
/assign @oomichi

oomichi commented on test/e2e/upgrades/topology_manager.go:
```diff
@@ -0,0 +1,317 @@
+	"k8s.io/apimachinery/pkg/util/uuid"
+
+	"k8s.io/client-go/kubernetes/scheme"
+
```
ditto

oomichi commented on test/e2e/upgrades/topology_manager.go:
```diff
@@ -0,0 +1,317 @@
+
+	"k8s.io/client-go/kubernetes/scheme"
+
+	kubeletconfigv1beta1 "k8s.io/kubelet/config/v1beta1"
```
A necessary blank line between `"k8s.io/*"` block except `"k8s.io/kubernetes"` and `"k8s.io/kubernetes"` block.

oomichi commented on test/e2e/upgrades/topology_manager.go:
```diff
@@ -0,0 +1,317 @@
+
+	kubeletconfigv1beta1 "k8s.io/kubelet/config/v1beta1"
+	kubeletconfig "k8s.io/kubernetes/pkg/kubelet/apis/config"
+
```
ditto: unnecessary blank line.

oomichi commented on test/e2e/upgrades/topology_manager.go:
```diff
@@ -0,0 +1,317 @@
+
+	"k8s.io/kubernetes/pkg/kubelet/cm/cpuset"
+	"k8s.io/kubernetes/pkg/kubelet/cm/topologymanager"
+
```
ditto: unnecessary blank line.

oomichi commented on test/e2e/upgrades/topology_manager.go:
```diff
@@ -0,0 +1,317 @@
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2ekubectl "k8s.io/kubernetes/test/e2e/framework/kubectl"
+	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+
```
ditto: unnecessary blank line.

oomichi commented on test/e2e/upgrades/topology_manager.go:
```diff
@@ -0,0 +1,317 @@
+	ginkgo.By("Verifying it is still possible to create a Pod with aligned resources")
+	node := getTopologyManagerSingleNumaNodePolicyEnabledNode(f)
+	if node == nil {
+		// XXX: but what if the topology manager WAS enabled before the upgrade?
```
What is this `XXX`? `NOTE` or something?

fromanirh replied:
was a `NOTE` to self which in retrospect is just misleading. Removed.

oomichi commented on test/e2e/upgrades/topology_manager.go:
```diff
@@ -0,0 +1,317 @@
+	framework.ExpectEqual(len(match), 2)
+	port, err := strconv.Atoi(match[1])
+	framework.ExpectNoError(err)
+	ginkgo.By("http requesting node kubelet /configz")
```
nit: better to put a new blank line before `ginkgo.By()` for code readability.

fromanirh replied:
done

fromanirh commented on test/e2e/upgrades/topology_manager.go:
```diff
@@ -0,0 +1,317 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/uuid"
+
```
I missed how can I check for these style violation on my box, saving review time. Could you please suggest a way (o r point me to the right docs :)) ?

ffromani committed:
e2e: upgrade: topology manager; review comments

Address review comments: add readability fixes.

Signed-off-by: Francesco Romani <fromani@redhat.com>

oomichi commented on test/e2e/upgrades/topology_manager.go:
```diff
@@ -0,0 +1,311 @@
+					Image: imageutils.GetE2EImage(imageutils.BusyBox),
+					Resources: v1.ResourceRequirements{
+						Requests: v1.ResourceList{
+							v1.ResourceName(v1.ResourceCPU):    resource.MustParse("2000m"),
```
Can I know why this pod requires 2 CPU cores?
That can be too much on some environments, I feel.

fromanirh replied:
Two cores is the minimum amount of resources we can use to actually check the resources are aligned. But here we want to check if the topology manager works correctly across upgrades, hence we maybe can simplify a bit more and require just one core. I'll do some research and put the PR on hold until then.

fromanirh replied:
The topology manager is actually exercised using just one core, but I can't find a way to test it worked as intended with less than two cores, so I need to keep it this way.

derekwaynecarr replied:
I think its fine that environments which require alignment testing verification during upgrades can reserve 2 cores.

oomichi commented on test/e2e/upgrades/topology_manager.go:
```diff
@@ -0,0 +1,311 @@
+			return fmt.Errorf("Unknown sibilings for cpu %d", cpuNum)
+		}
+
+		if !reflect.DeepEqual(sibilingsPrev, sibilings) {
```
Does this require that all CPU cores should be in the same siblings from a pod viewpoint?
BTW typo: s/sibilings/siblings/

fromanirh replied:
fixed typo and clarified the intent of the test.

oomichi reviewed (commented):
Thanks for updating.
Let's test anyways.

/ok-to-test

ffromani committed:
e2e: upgrade: clarify test intent and fix typos

Add a comment to explain the meaning the test we perform,
and fix the sibiling -> sibling typo.

Signed-off-by: Francesco Romani <fromani@redhat.com>

fromanirh commented:
/hold
need to see if we can reduce the resource request to just 1 core

fromanirh commented:
/hold cancel
it seems two full cores is really the minimum we need to test the topology manager is working

ffromani committed:
e2e: upgrade: run proxy just once

refactor further the code to avoid to run a proxy instance
for each call.

We don't really need to stop and start the proxy while checking each
worker node, and doing it just once makes the test a bit faster
and a bit less fragile (less moving parts)

The refactoring done in this patch allows us to run the proxy just once.

Signed-off-by: Francesco Romani <fromani@redhat.com>

derekwaynecarr reviewed (approved):
this looks good from sig-node perspective.

/approve
/lgtm

oomichi commented:
Thanks for updating.

/lgtm

fromanirh commented:
/hold

cynepco3hahue commented:
@derekwaynecarr @oomichi Hi guys, I can see that we skip upgrade tests under the `pull-kubernetes-e2e-gce`, maybe do we have some CI lane that can be initiated manually to run upgrade tests?

fromanirh commented:
I tried to run the e2e upgrade flow on my box on gcp with no success yet, do we have up to date instructions?

dchen1107 commented:
/lgtm
/approve

dashpole commented:
/cc

fromanirh commented:
> I tried to run the e2e upgrade flow on my box on gcp with no success yet, do we have up to date instructions?

Looks like it works now: I was just lacking a couple of gsutil/gcloud commands in the steps I followed. I'll update as soon as I have some progress.

ffromani committed:
e2e: upgrade: topomgr: minimize the requirements

The GCP environment is oftentimes too resource constrained.
Worker nodes can have as little as 2 cpus, hence we cannot
require two full cores to actually check the topology manager behaviour,
because the PODs will fail to schedule (OutOfCPU), leading to
false negatives.

Hence, we narrow down the tests to the bare minimum: trigger
the topology manager flow, and check the pod actually runs.
To save resources, we delete it right after.

For the same reasons, we isolate the upgrade tests in their own spec,
to make sure they interfer as little as possible with the
existing tests, and to make it possible to run them only if needed.

Signed-off-by: Francesco Romani <fromani@redhat.com>

fromanirh commented:
how come the approved label wasn't removed? I think we need another review, the last changes are pretty significant.

oomichi commented:
@fromanirh 
You need to update this PR with `./hack/update-bazel.sh` to pass pull-kubernetes-verify test job.

fromanirh commented:
/retest

oomichi commented on test/e2e/cloud/gcp/cluster_upgrade.go:
```diff
@@ -279,6 +283,72 @@ var _ = SIGDescribe("gpu Upgrade [Feature:GPUUpgrade]", func() {
 	})
 })
 
+var _ = SIGDescribe("TopologyManager Upgrade [Feature:TopologyManagerUpgrade]", func() {
```
At this time, test-infra contains just `"\[Feature:TopologyManager\]"` without `Upgrade` keyword in sig-node/node-kubelet.yaml and sig-node/sig-node-presubmit.yaml.
Will you add the above keyword also? Or should we change here to `"\[Feature:TopologyManager\]"` to run this test on the above test jobs?

fromanirh replied:
Will sync with the node team and figure out, I don't have an answer for this.

fromanirh replied:
Hi, partial followup, still figuring out the path forward with the other parties.
I was trying to make my code work and eventually checked out how the existing code is working, but according to https://testgrid.k8s.io/google-gce-upgrade even the existing tests are red, and the last upgrade check configured was from 1.13 and 1.14. This was unexpected, am I looking in the wrong place perhaps? If so what's the right place to check how existing test are doing?
If the existing tests are indeed red the simplest way forward is drop this PR and restart from scratch in the node area.
Thanks for any help.

oomichi reviewed (commented):
@cynepco3hahue 

> @derekwaynecarr @oomichi Hi guys, I can see that we skip upgrade tests under the pull-kubernetes-e2e-gce, maybe do we have some CI lane that can be initiated manually to run upgrade tests?

That is a nice point.
The test contains `[Feature:***]` label which tests are skipped in general test jobs as  https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/88107/pull-kubernetes-e2e-gce/1233340232040976384/build-log.txt which shows `--ginkgo.skip=\[Slow\]|\[Serial\]|\[Disruptive\]|\[Flaky\]|\[Feature:.+\]` is specified.
We can see the test result on different test jobs which are specific SIG ones.

According to the above comment, I have a question about this test job.
Please check my inline comment.

fromanirh commented:
@oomichi ack, I've seen your question but I need to sync in order to provide an answer. Will respond ASAP.

Meanwhile: I updated the tests and squashed a few little bugs in the process. Yet, they still don't complete on my env. However also a clean master snapshot doesn't complete for me. Here's what I did:

k8s master
	fdb2cb4c883
build using
```kubetest --build --stage=gs://ABC --check-version-skew=false --gcp-project=XXX```
INTENTIONALLY made all objects in the bucket world-readable (AFAIU the e2e tests require this, but this is just a nightly build so nothing to hide):
```gsutil -m acl -r ch -u AllUsers:R gs://ABC```
run using
``kubetest --up --test --check-version-skew=false --extract=gs://ABC/v1.19.0-alpha.0.399+fdb2cb4c8832da --extract=v1.17.2 --upgrade_args=--ginkgo.focus="Feature:Upgrade" --gcp-project=XXX --down```

I narrowed down to this: when the test run cluster/gce/upgrade.sh, the test get stuck there

```
  echo "== Waiting for new node to be added to k8s.  ==" >&2
  while true; do
    local new_boot_id=$("${KUBE_ROOT}/cluster/kubectl.sh" get node "${instance}" --output=jsonpath='{.status.nodeInfo.bootID}' 2>&1) && kubectl_rc=$? || kubectl_rc=$?
    if [[ "${kubectl_rc}" != 0 ]]; then
      echo "== FAILED to get node ${instance} =="
      echo "${boot_id}"
      echo "  (Will retry.)"
    elif [[ "${boot_id}" != "${new_boot_id}" ]]; then
      echo "Node ${instance} recreated."
      break
    else
      echo -n .
    fi
    sleep 1
  done
```
it seems the `bootID` never changes. I waited for 30m with no changes. Maybe I just need to wait more? 

Any suggestion to make it move forward is welcome.

fromanirh commented:
quick update: I had to work on other stuff, but I definitely wants to understand what's wrong there and I should be able to work on this again soon. Thanks.

oomichi commented:
@fromanirh 

> @oomichi ack, I've seen your question but I need to sync in order to provide an answer. Will respond ASAP.

I am just waiting for your answer for the above.

fromanirh commented:
> @fromanirh
> 
> > @oomichi ack, I've seen your question but I need to sync in order to provide an answer. Will respond ASAP.
> 
> I am just waiting for your answer for the above.

Absolutely, sorry for the delay. I don't have yet an answer, need to get in touch with relevant people.  Hopefully this will happen next week. 
What I meant in my previous comment is that I still have some investigation to do on my side (hence the PR still on hold)

fromanirh commented:
considering the work begun on sig-node to revamp and clean up our e2e tests, I'm closing this PR. We'll evaluate how to introduce a new upgrade test there.
Thanks to all the reviewers.
